lib/chibi/syntax-case.scm

```diff
@@ -0,0 +1,323 @@
+;; Written by Marc Nieper-Wißkirchen
+
+;; TODO: make-variable-transformer and identifier-syntax.
+
+;; TODO: make-synthetic-identifier should return a truly unique (that
+;; is not free-identifier=? to any other) identifier.
+
+;; TODO: Consecutive ellipses in syntax templates.
+
+;; TODO: Write many more tests.
+
+(define-syntax define-pattern-variable
+  (er-macro-transformer
+   (lambda (expr rename compare)
+     (let ((id (cadr expr))
+	   (binding (cddr expr)))
+       (let ((mac (cdr (env-cell (current-usage-environment) id))))
+	 (macro-aux-set! mac binding))
+       `(,(rename 'begin))))))
+
+(define (make-pattern-variable pvar)
+  (lambda (expr)
+    (error "reference to pattern variable outside syntax" pvar)))
+
+(define (pattern-variable x)
+  (let ((cell (env-cell (current-usage-environment) x)))
+    (and cell (macro? (cdr cell)) (macro-aux (cdr cell)))))
+
+(define (rename id)
+  ((current-renamer) id))
+
+(define (ellipsis? id)
+  (free-identifier=? id (rename '...)))
+
+(define bound-identifier=?
+  (lambda (x y)
+    (eq? x y)))
+
+(define (syntax-transformer level)
+  (er-macro-transformer
+   (lambda (expr rename compare)
+     (let*-values (((out envs)
+		    (gen-template (cadr expr) '() ellipsis? level)))
+       out))))
+
+(define (syntax->datum stx)
+  (strip-syntactic-closures stx))
+
+(define (symbol->identifier id symbol)
+  (if (symbol? id)
+      symbol
+      ((syntactic-closure-rename id)
```

Comment:
This can be false with sc- and rsc- macros, right?

Reply:
I don't see how to implement `datum->syntax` for sc- and rsc-macros. I'm not even sure how compatible they are to syntax-rules macros in the case of unhygienic macros that need `datum->syntax` on the syntax-rules/er-macro-transformer side.

Thus in `symbol->identifier` the identifier `id` (if it is not a symbol) has to be an identifier created by the renaming procedure of `er-macro-transformer` or of the form `#'x` (where `x` is a syntactic identifier), which is basically the same.

General identifiers created by `make-syntactic-closure` simply do not contain enough information about their syntactic environment. `symbol->identifier` could either raise an error in this case or return `symbol` unchanged.

Reply:
How about we warn and create a new identifier with the `current-renamer`?

Reply:
What do you think of my idea (which I wrote about yesterday in Chibi's newsgroup) to attach the renamer not to the syntactic closure representing an identifier but to the environment (which is stored inside the syntactic closure)? In that case, all transformer environments would bundle their (non-trivial) renamer. For all other environments, the identity renamer would be the default.